content/client-lib-development-guide/features.textile

```diff
@@ -488,6 +490,11 @@ h3(#realtime-channel). Channel
 ** @(RTL4f)@ Once an @ATTACH@ @ProtocolMessage@ is sent, if an @ATTACHED@ @ProtocolMessage@ is not received within the "default realtime request timeout":#defaults, the attach request should be treated as though it has failed and the channel should transition to the @SUSPENDED@ state. The channel will then be subsequently automatically re-attached as described in "RTL13":#RTL13
 ** @(RTL4d)@ If the language permits, a callback can be provided that is called when the channel is attached successfully or the attach fails and the @ErrorInfo@ error is passed as an argument to the callback
 ** @(RTL4e)@ If the user does not have sufficient permissions to attach to the channel, the channel will transition to @FAILED@ and set the @Channel#errorReason@
+** @(RTL4j)@ If the attach is not a clean attach (where a 'clean attach' is an attach attempt where the channel has either not previously been attached or has been explicitly detached since the last time it was attached); for example an automatic reattach triggered by "@RTN15c3@":#RTN15c3 or "@RTL13a@":#RTL13a (non-exhaustive); the library should set the "@ATTACH_RESUME@":#TR3f flag in the @ATTACH@ message
+** @(RTL4k)@ If the user has specified a non-empty @params@ object in the @ChannelOptions@ ("@TB2c@":#TB2c), it must be included in a @params@ field of the @ATTACH@ @ProtocolMessage@
+*** @(RTL4k1)@ If @params@ are sent in the @ATTACH@ message, the corresponding @ATTACHED@ (and any subsequent @ATTACHED@s) will contain the subset of those params that the server recognises and validates. This should be exposed in a read-only @params@ field of the @Channel@ (or a @getParams()@ method where that is more idiomatic)
+** @(RTL4l)@ If the user has specified a @modes@ array in the @ChannelOptions@ ("@TB2d@":#TB2d), it must be encoded as a bitfield per "@TR3@":#TR3 and set as the @flags@ field of the @ATTACH@ @ProtocolMessage@. (For the avoidance of doubt, when multiple different spec items require flags to be set in the @ATTACH@, the final @flags@ field should be the bitwise OR of them all)
```

Comment:
What would happen in a language where `modes` is an empty array as opposed to a `null` array.  Is an empty array or `null` value ignored in both these cases?  Should we be specific about that?

Reply:
Makes no difference -- encoding an empty array into a bitfield is equivalent to bitwise-ORing the bitfield with 0, which is a no-op.

Realtime behaviour after https://github.com/ably/realtime/pull/2741/files#diff-0757a77e03530e3b881b82269acaa354R2305 is merged will be to apply the default set of modes if the flags field does not have any modes set (a change we had to make as a consequence of adding the new M_ATTACH_RESUME flag since it's then no longer true that the client only includes the flags field if it wants to specify modes)